README.md
First Data Latvia gateway for Active Merchant
=============================================

[![Build Status](https://travis-ci.org/ebeigarts/active_merchant_first_data.svg?branch=master)](https://travis-ci.org/ebeigarts/active_merchant_first_data)

## Install

```bash
$ gem install active_merchant_first_data
```

## Usage

```ruby
require "active_merchant_first_data"

gateway = ActiveMerchant::Billing::FirstDataGateway.new(
  pem:  File.read("1234567_keystore.pem"),
  pem_password: "5x64jq8n234c"
)

# Authorize for 10 euros (1000 euro cents)
response = gateway.authorize(1000, client_ip_addr: '127.0.0.1')

# Use this url to enter credit card data
gateway.redirect_url(response.authorization)

# Capture the money
gateway.capture(1000, response.authorization, client_ip_addr: '127.0.0.1')
```

## First Data test environment setup

1. Generate a new certificate

    ```bash
    $ openssl req -newkey rsa:2048 -keyout spec/certs/1234567_key.pem -out spec/certs/1234567_req.pem -subj "/C=lv/O=example.com/CN=1234567" -outform PEM
    Enter PEM pass phrase: 81f174259v45
    ```

2. [Request your certificate using `1234567_req.pem`](https://secureshop-test.firstdata.lv/report/keystore_.do)

3. Copy the 3 files you received in e-mail to `spec/certs/`:

    ```
    1234567.pem
    1234567_certificate_chain.p7.pem
    ECOMM.pem
    ```

4. Convert the certificates and keys to `1234567_keystore.pem`

    ```bash
    $ openssl pkcs12 -export -in spec/certs/1234567.pem -out spec/certs/1234567_keystore.p12 -certfile spec/certs/ECOMM.pem -inkey spec/certs/1234567_key.pem
    Enter pass phrase for 1234567_key.pem: 81f174259v45
    Enter Export Password: <empty>
    ```

    ```bash
    $ openssl pkcs12 -in spec/certs/1234567_keystore.p12 > spec/certs/1234567_keystore.pem
    Enter Import Password: <empty>
    Enter PEM pass phrase: 5x64jq8n234c
    ```

5. [Set your WAN IP address](https://secureshop-test.firstdata.lv/report/merchantlist.do)


CHANGELOG.md
## 1.3.1

*   Added AZN and RON currencies (@EdgeLimits, #8)

## 1.3.0

*   Response as an object instead of a hash (@Sacristan, #6)

## 1.2.0

*   Changed default currency from LVL to EUR

## 1.1.0

*   Recurring payments support (@rsim)

## 1.0.0

*   Initial version


spec/spec_helper.rb
$:.unshift File.dirname(__FILE__) + '/../lib'

require "rubygems"
require "bundler/setup"
require "active_merchant_first_data"
require "logger"
require "vcr"

VCR.config do |c|
  c.cassette_library_dir = "spec/cassettes"
  c.stub_with :fakeweb
  # c.default_cassette_options = { :record => :new_episodes }
end

RSpec.configure do |config|
  config.extend VCR::RSpec::Macros
end

ActiveMerchant::Billing::Base.gateway_mode = :test
ActiveMerchant::Billing::FirstDataGateway.ssl_strict = true
ActiveMerchant::Billing::FirstDataGateway.logger = Logger.new(STDOUT)
ActiveMerchant::Billing::FirstDataGateway.logger.level = Logger::WARN
# ActiveMerchant::Billing::FirstDataGateway.wiredump_device = STDOUT


lib/active_merchant/billing/first_data/gateway.rb
module ActiveMerchant #:nodoc:
  module Billing #:nodoc:
    module FirstData
      class Gateway < Gateway
        class_attribute :test_url, :live_url, :test_redirect_url, :live_redirect_url, :pem_file, :pem_password

        self.test_redirect_url = 'https://secureshop-test.firstdata.lv/ecomm/ClientHandler'
        self.live_redirect_url = 'https://secureshop.firstdata.lv/ecomm/ClientHandler'

        self.test_url = 'https://secureshop-test.firstdata.lv:8443/ecomm/MerchantHandler'
        self.live_url = 'https://secureshop.firstdata.lv:8443/ecomm/MerchantHandler'

        self.homepage_url = 'http://www.firstdata.lv/'
        self.display_name = 'First Data'

        self.ssl_strict = false
        self.default_currency = '978' # EUR (http://en.wikipedia.org/wiki/ISO_4217)
        self.money_format = :cents

        CURRENCY_CODES = {
          'ADP' => '020', 'AED' => '784', 'AFA' => '004', 'ALL' => '008', 'AMD' => '051',
          'ANG' => '532', 'AOA' => '973', 'ARS' => '032', 'AUD' => '036', 'AWG' => '533',
          'AZM' => '031', 'BAM' => '977', 'BBD' => '052', 'BDT' => '050', 'BGL' => '100',
          'BGN' => '975', 'BHD' => '048', 'BIF' => '108', 'BMD' => '060', 'BND' => '096',
          'BOB' => '068', 'BOV' => '984', 'BRL' => '986', 'BSD' => '044', 'BTN' => '064',
          'BWP' => '072', 'BYR' => '974', 'BZD' => '084', 'CAD' => '124', 'CDF' => '976',
          'CHF' => '756', 'CLF' => '990', 'CLP' => '152', 'CNY' => '156', 'COP' => '170',
          'CRC' => '188', 'CUP' => '192', 'CVE' => '132', 'CYP' => '196', 'CZK' => '203',
          'DJF' => '262', 'DKK' => '208', 'DOP' => '214', 'DZD' => '012', 'ECS' => '218',
          'ECV' => '983', 'EEK' => '233', 'EGP' => '818', 'ERN' => '232', 'ETB' => '230',
          'EUR' => '978', 'FJD' => '242', 'FKP' => '238', 'GBP' => '826', 'GEL' => '981',
          'GHC' => '288', 'GIP' => '292', 'GMD' => '270', 'GNF' => '324', 'GTQ' => '320',
          'GWP' => '624', 'GYD' => '328', 'HKD' => '344', 'HNL' => '340', 'HRK' => '191',
          'HTG' => '332', 'HUF' => '348', 'IDR' => '360', 'ILS' => '376', 'INR' => '356',
          'IQD' => '368', 'IRR' => '364', 'ISK' => '352', 'JMD' => '388', 'JOD' => '400',
          'JPY' => '392', 'KES' => '404', 'KGS' => '417', 'KHR' => '116', 'KMF' => '174',
          'KPW' => '408', 'KRW' => '410', 'KWD' => '414', 'KYD' => '136', 'KZT' => '398',
          'LAK' => '418', 'LBP' => '422', 'LKR' => '144', 'LRD' => '430', 'LSL' => '426',
          'LTL' => '440', 'LVL' => '428', 'LYD' => '434', 'MAD' => '504', 'MDL' => '498',
          'MGF' => '450', 'MKD' => '807', 'MMK' => '104', 'MNT' => '496', 'MOP' => '446',
          'MRO' => '478', 'MTL' => '470', 'MUR' => '480', 'MVR' => '462', 'MWK' => '454',
          'MXN' => '484', 'MXV' => '979', 'MYR' => '458', 'MZM' => '508', 'NAD' => '516',
          'NGN' => '566', 'NIO' => '558', 'NOK' => '578', 'NPR' => '524', 'NZD' => '554',
          'OMR' => '512', 'PAB' => '590', 'PEN' => '604', 'PGK' => '598', 'PHP' => '608',
          'PKR' => '586', 'PLN' => '985', 'PYG' => '600', 'QAR' => '634', 'ROL' => '642',
          'RUB' => '643', 'RUR' => '810', 'RWF' => '646', 'SAR' => '682', 'SBD' => '090',
          'SCR' => '690', 'SDD' => '736', 'SEK' => '752', 'SGD' => '702', 'SHP' => '654',
          'SIT' => '705', 'SKK' => '703', 'SLL' => '694', 'SOS' => '706', 'SRG' => '740',
          'STD' => '678', 'SVC' => '222', 'SYP' => '760', 'SZL' => '748', 'THB' => '764',
          'TJS' => '972', 'TMM' => '795', 'TND' => '788', 'TOP' => '776', 'TPE' => '626',
          'TRL' => '792', 'TRY' => '949', 'TTD' => '780', 'TWD' => '901', 'TZS' => '834',
          'UAH' => '980', 'UGX' => '800', 'USD' => '840', 'UYU' => '858', 'UZS' => '860',
          'VEB' => '862', 'VND' => '704', 'VUV' => '548', 'XAF' => '950', 'XCD' => '951',
          'XOF' => '952', 'XPF' => '953', 'YER' => '886', 'YUM' => '891', 'ZAR' => '710',
          'ZMK' => '894', 'ZWD' => '716'
        }

        class Error < StandardError
          attr_reader :response

          def initialize(response, msg=nil)
            @response = response
            super(msg || response[:error])
          end
        end

        # Creates a new FirstDataGateway
        #
        # The gateway requires that a valid pem and password be passed
        # in the +options+ hash.
        #
        # @option options [String] :pem First Data cert/key (REQUIRED)
        # @option options [String] :pem_password First Data cert password (REQUIRED)
        #
        def initialize(options = {})
          requires!(options, :pem, :pem_password)
          @options = options
          super
        end

        # Perform a purchase, which is essentially an authorization and capture in a single operation.
        #
        # Registering of SMS transaction
        #
        # @param [Integer] amount transaction amount in minor units, mandatory
        # @option params [Integer, String] :currency transaction currency code, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        # @option params [String] :description description of transaction, optional
        # @option params [String] :language authorization language identificator, optional
        # @return [ActiveSupport::HashWithIndifferentAccess] TRANSACTION_ID
        #
        def purchase(amount, params = {})
          params = params.reverse_merge(
            :command => :v,
            :amount => amount,
            :currency => default_currency
          )
          lookup_currency(params)
          requires!(params, :amount, :currency, :client_ip_addr)
          commit(params)
        end

        # Performs an authorization, which reserves the funds on the customer's credit card, but does not
        # charge the card.
        #
        # Registering of DMS authorisation
        #
        # @param [Integer] amount transaction amount in minor units, mandatory
        # @option params [Integer, String] :currency transaction currency code, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        # @option params [String] :description description of transaction, optional
        # @option params [String] :language authorization language identificator, optional
        # @return [ActiveSupport::HashWithIndifferentAccess] TRANSACTION_ID
        #
        def authorize(amount, params = {})
          params = params.reverse_merge(
            :command => :a,
            :msg_type => 'DMS',
            :amount => amount,
            :currency => default_currency
          )
          lookup_currency(params)
          requires!(params, :amount, :currency, :client_ip_addr, :msg_type)
          commit(params)
        end

        # Captures the funds from an authorized transaction.
        #
        # Making of DMS transaction
        #
        # @param [Integer] amount transaction amount in minor units, mandatory
        # @param [Integer] trans_id id of previously made successeful authorisation
        # @option params [Integer, String] :currency transaction currency code, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        # @option params [String] :description description of transaction, optional
        # @return [ActiveSupport::HashWithIndifferentAccess] RESULT, RESULT_CODE, RRN, APPROVAL_CODE
        #
        def capture(amount, trans_id, params = {})
          params = params.reverse_merge(
            :command => :t,
            :msg_type => 'DMS',
            :trans_id => trans_id,
            :amount => amount,
            :currency => default_currency
          )
          lookup_currency(params)
          requires!(params, :trans_id, :amount, :currency, :client_ip_addr)
          commit(params)
        end

        # Transaction result
        #
        # @param [Integer] trans_id transaction identifier, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        # @return [ActiveSupport::HashWithIndifferentAccess] RESULT, RESULT_CODE, 3DSECURE, AAV, RRN, APPROVAL_CODE
        #
        def result(trans_id, params = {})
          params = params.reverse_merge(
            :command => :c,
            :trans_id => trans_id
          )
          requires!(params, :trans_id, :client_ip_addr)
          commit(params)
        end

        # refund() allows you to return money to a card that was previously billed.
        #
        # Transaction reversal
        #
        # @param [Integer] amount transaction amount in minor units, mandatory
        # @param [Integer] trans_id transaction identifier, mandatory
        # @return [ActiveSupport::HashWithIndifferentAccess] RESULT, RESULT_CODE
        #
        def refund(amount, trans_id = nil)
          params = {
            :command => :r,
            :trans_id => trans_id,
            :amount => amount
          }
          requires!(params, :command, :trans_id, :amount)
          commit(params)
        end

        def credit(amount, trans_id = nil)
          deprecated CREDIT_DEPRECATION_MESSAGE
          refund(amount, trans_id)
        end

        # Register new recurring payment along with the first payment
        #
        # @param [Integer] amount transaction amount in minor units, mandatory (up to 12 digits)
        # @option params [Integer, String] :currency transaction currency code, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        # @option params [String] :description transaction details, mandatory (up to 125 characters)
        # @option params [String] :biller_client_id recurring payment identifier, mandatory (up to 30 characters)
        # @option params [String] :perspayee_expiry preferred deadline for a Recurring payment, mandatory (MMYY)
        #   (system will compare two expiry dates - recurring payment expiry date provided by Merchant and
        #   card expiry date provided by Cardholder. In response Merchant will receive back the earliest expiry dates of both).
        #   For example, Recurring payment date is 1214 but card expiry date 1213 then Merchant will receive back 1213.
        #   First Data system will save the earliest expiry date as Recurring payment expiry date.
        # @option params [Integer] :perspayee_overwrite set to 1 to overwrite existing recurring payment card data
        #   together with payment, optional
        # @option params [String] :language authorization language identificator, optional
        # @return [ActiveSupport::HashWithIndifferentAccess] TRANSACTION_ID
        #
        # Afterwards when transaction result is requested then result response includes also RECC_PMNT_ID and RECC_PMNT_EXPIRY
        #
        def recurring(amount, params = {})
          params = params.reverse_merge(
            :command => :z,
            :amount => amount,
            :currency => default_currency,
            :msg_type => 'SMS',
            :perspayee_gen => 1
          )
          lookup_currency(params)
          requires!(params, :amount, :currency, :client_ip_addr, :description, :biller_client_id, :perspayee_expiry)
          commit(params)
        end

        # Execute subsequent recurring payment
        #
        # @param [Integer] amount transaction amount in minor units, mandatory (up to 12 digits)
        # @option params [Integer, String] :currency transaction currency code, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        #   (the same IP address which was provided when registered recurring payment)
        # @option params [String] :description transaction details, mandatory (up to 125 characters)
        # @option params [String] :biller_client_id recurring payment identifier, mandatory (up to 30 characters)
        # @return [ActiveSupport::HashWithIndifferentAccess] TRANSACTION_ID, RESULT, RESULT_CODE, RRN, APPROVAL_CODE
        #
        def execute_recurring(amount, params = {})
          params = params.reverse_merge(
            :command => :e,
            :amount => amount,
            :currency => default_currency
          )
          lookup_currency(params)
          requires!(params, :amount, :currency, :client_ip_addr, :description, :biller_client_id)
          commit(params)
        end

        # Overwriting existing recurring payment card data without payment
        #
        # @option params [Integer, String] :currency transaction currency code, mandatory
        # @option params [String] :client_ip_addr client's IP address, mandatory
        # @option params [String] :description transaction details, mandatory (up to 125 characters)
        # @option params [String] :biller_client_id existing recurring payment identifier, mandatory (up to 30 characters)
        # @option params [String] :perspayee_expiry preferred deadline for a Recurring payment, mandatory (MMYY)
        #   (system will compare two expiry dates - recurring payment expiry date provided by Merchant and
        #   card expiry date provided by Cardholder. In response Merchant will receive back the earliest expiry dates of both).
        #   For example, Recurring payment date is 1214 but card expiry date 1213 then Merchant will receive back 1213.
        #   First Data system will save the earliest expiry date as Recurring payment expiry date.
        # @option params [String] :language authorization language identificator, optional
        # @return [ActiveSupport::HashWithIndifferentAccess] TRANSACTION_ID
        #
        # Afterwards when transaction result is requested then result response includes also RECC_PMNT_ID and RECC_PMNT_EXPIRY
        #
        def update_recurring(params = {})
          params = params.reverse_merge(
            :command => :p,
            :amount => 0,
            :currency => default_currency,
            :msg_type => 'AUTH',
            :perspayee_gen => 1,
            :perspayee_overwrite => 1
          )
          lookup_currency(params)
          requires!(params, :currency, :client_ip_addr, :description, :biller_client_id, :perspayee_expiry)
          commit(params)
        end

        # Close business day.
        def close_day
          commit({ :command => :b })
        end

        def endpoint_url
          test? ? test_url : live_url
        end

        def redirect_url(trans_id = nil)
          url = test? ? test_redirect_url : live_redirect_url
          url += "?trans_id=#{CGI.escape trans_id}" if trans_id
          url
        end

        private
          def lookup_currency(params)
            params[:currency] = CURRENCY_CODES[params[:currency]] || params[:currency]
          end

          # Convert HTTP response body to a Ruby Hash.
          def parse(body)
            results = ActiveSupport::HashWithIndifferentAccess.new
            body.split(/[\r\n]+/).each do |pair|
              key, val = pair.split(": ")
              results[key.downcase] = val
            end
            results
          end

          def commit(params = {})
            response = parse(ssl_post(endpoint_url, post_data(params)))
            # FIXME: test cases 17 and 19 return unnecessary error even when result and result_code are present
            # should be removed when this issue is fixed on gataway side
            raise Error.new(response) if !response[:error].blank? && response[:result_code].blank?
            ActiveMerchant::Billing::FirstData::Response.new response
          end

          def post_data(params)
            post = PostData.new
            params.each { |k, v| post[k] = v }
            post.to_s
          end
      end
    end
  end
end


lib/active_merchant/billing/first_data/response.rb
module ActiveMerchant #:nodoc:
  module Billing #:nodoc:
    module FirstData
      class Response

        FDL_RESPONSE_CODES = {
          c000: "Approved",
          c001: "Approved, honour with identification",
          c002: "Approved for partial amount",
          c003: "Approved for VIP",
          c004: "Approved, update track 3",
          c005: "Approved, account type specified by card issuer",
          c006: "Approved for partial amount, account type specified by card issuer",
          c007: "Approved, update ICC",
          c100: "Decline (general, no comments)",
          c101: "Decline, expired card",
          c102: "Decline, suspected fraud",
          c103: "Decline, card acceptor contact acquirer",
          c104: "Decline, restricted card",
          c105: "Decline, card acceptor call acquirer's security department",
          c106: "Decline, allowable PIN tries exceeded",
          c107: "Decline, refer to card issuer",
          c108: "Decline, refer to card issuer's special conditions",
          c109: "Decline, invalid merchant",
          c110: "Decline, invalid amount",
          c111: "Decline, invalid card number",
          c112: "Decline, PIN data required",
          c113: "Decline, unacceptable fee",
          c114: "Decline, no account of type requested",
          c115: "Decline, requested function not supported",
          c116: "Decline, not sufficient funds",
          c117: "Decline, incorrect PIN",
          c118: "Decline, no card record",
          c119: "Decline, transaction not permitted to cardholder",
          c120: "Decline, transaction not permitted to terminal",
          c121: "Decline, exceeds withdrawal amount limit",
          c122: "Decline, security violation",
          c123: "Decline, exceeds withdrawal frequency limit",
          c124: "Decline, violation of law",
          c125: "Decline, card not effective",
          c126: "Decline, invalid PIN block",
          c127: "Decline, PIN length error",
          c128: "Decline, PIN kay synch error",
          c129: "Decline, suspected counterfeit card",
          c200: "Pick-up (general, no comments)",
          c201: "Pick-up, expired card",
          c202: "Pick-up, suspected fraud",
          c203: "Pick-up, card acceptor contact card acquirer",
          c204: "Pick-up, restricted card",
          c205: "Pick-up, card acceptor call acquirer's security department",
          c206: "Pick-up, allowable PIN tries exceeded",
          c207: "Pick-up, special conditions",
          c208: "Pick-up, lost card",
          c209: "Pick-up, stolen card",
          c210: "Pick-up, suspected counterfeit card",
          c300: "Status message: file action successful",
          c301: "Status message: file action not supported by receiver",
          c302: "Status message: unable to locate record on file",
          c303: "Status message: duplicate record, old record replaced",
          c304: "Status message: file record field edit error",
          c305: "Status message: file locked out",
          c306: "Status message: file action not successful",
          c307: "Status message: file data format error",
          c308: "Status message: duplicate record, new record rejected",
          c309: "Status message: unknown file",
          c400: "Accepted (for reversal)",
          c500: "Status message: reconciled, in balance",
          c501: "Status message: reconciled, out of balance",
          c502: "Status message: amount not reconciled, totals provided",
          c503: "Status message: totals for reconciliation not available",
          c504: "Status message: not reconciled, totals provided",
          c600: "Accepted (for administrative info)",
          c601: "Status message: impossible to trace back original transaction",
          c602: "Status message: invalid transaction reference number",
          c603: "Status message: reference number/PAN incompatible",
          c604: "Status message: POS photograph is not available",
          c605: "Status message: requested item supplied",
          c606: "Status message: request cannot be fulfilled - required documentation is not available",
          c700: "Accepted (for fee collection)",
          c800: "Accepted (for network management)",
          c900: "Advice acknowledged, no financial liability accepted",
          c901: "Advice acknowledged, finansial liability accepted",
          c902: "Decline reason message: invalid transaction",
          c903: "Status message: re-enter transaction",
          c904: "Decline reason message: format error",
          c905: "Decline reason message: acqiurer not supported by switch",
          c906: "Decline reason message: cutover in process",
          c907: "Decline reason message: card issuer or switch inoperative",
          c908: "Decline reason message: transaction destination cannot be found for routing",
          c909: "Decline reason message: system malfunction",
          c910: "Decline reason message: card issuer signed off",
          c911: "Decline reason message: card issuer timed out",
          c912: "Decline reason message: card issuer unavailable",
          c913: "Decline reason message: duplicate transmission",
          c914: "Decline reason message: not able to trace back to original transaction",
          c915: "Decline reason message: reconciliation cutover or checkpoint error",
          c916: "Decline reason message: MAC incorrect",
          c917: "Decline reason message: MAC key sync error",
          c918: "Decline reason message: no communication keys available for use",
          c919: "Decline reason message: encryption key sync error",
          c920: "Decline reason message: security software/hardware error - try again",
          c921: "Decline reason message: security software/hardware error - no action",
          c922: "Decline reason message: message number out of sequence",
          c923: "Status message: request in progress",
          c940: "Decline, blocked by fraud filter",
          c950: "Decline reason message: violation of business arrangement",
          c198: "Declined, call Card Processing Centre",
          c197: "Declined, call AmEx"
        }

        def initialize response_parameters
          self.parameters=response_parameters
        end

        def ok?
          result == "OK"
        end

        def success?
          ok?
        end

        def recurring?
          recc_pmnt_id.present? && recc_pmnt_expiry.present?
        end

        def result
          parameters[:result]
        end

        def authorization
          transaction_id
        end

        def transaction_id
          parameters[:transaction_id]
        end

        def result_code
          parameters[:result_code]
        end

        def recc_pmnt_id
          parameters[:recc_pmnt_id]
        end

        def recc_pmnt_expiry
          parameters[:recc_pmnt_expiry]
        end

        def three_dee_secure
          parameters[:'3dsecure']
        end

        def message
          result_text
        end

        def result_message
          result_text
        end

        def result_text
          FDL_RESPONSE_CODES["c#{result_code}".to_sym]
        end

        def []
          parameters
        end

        def [](value)
          parameters[value]
        end

        def parameters
          @parameters
        end

        private
          def parameters=(value)
            @parameters=value
          end
      end
    end
  end
end


lib/active_merchant/billing/first_data_response.rb
module ActiveMerchant #:nodoc:
  module Billing #:nodoc:
    class FirstDataResponse < Response
      RESPONSE_CODES = {
        "000" => "Approved",
        "001" => "Approved, honour with identification",
        "002" => "Approved for partial amount",
        "003" => "Approved for VIP",
        "004" => "Approved, update track 3",
        "005" => "Approved, account type specified by card issuer",
        "006" => "Approved for partial amount, account type specified by card issuer",
        "007" => "Approved, update ICC",
        "100" => "Decline (general, no comments)",
        "101" => "Decline, expired card",
        "102" => "Decline, suspected fraud",
        "103" => "Decline, card acceptor contact acquirer",
        "104" => "Decline, restricted card",
        "105" => "Decline, card acceptor call acquirer's security department",
        "106" => "Decline, allowable PIN tries exceeded",
        "107" => "Decline, refer to card issuer",
        "108" => "Decline, refer to card issuer's special conditions",
        "109" => "Decline, invalid merchant",
        "110" => "Decline, invalid amount",
        "111" => "Decline, invalid card number",
        "112" => "Decline, PIN data required",
        "113" => "Decline, unacceptable fee",
        "114" => "Decline, no account of type requested",
        "115" => "Decline, requested function not supported",
        "116" => "Decline, not sufficient funds",
        "117" => "Decline, incorrect PIN",
        "118" => "Decline, no card record",
        "119" => "Decline, transaction not permitted to cardholder",
        "120" => "Decline, transaction not permitted to terminal",
        "121" => "Decline, exceeds withdrawal amount limit",
        "122" => "Decline, security violation",
        "123" => "Decline, exceeds withdrawal frequency limit",
        "124" => "Decline, violation of law",
        "125" => "Decline, card not effective",
        "126" => "Decline, invalid PIN block",
        "127" => "Decline, PIN length error",
        "128" => "Decline, PIN kay synch error",
        "129" => "Decline, suspected counterfeit card",
        "197" => "Declined, call AmEx",
        "198" => "Declined, call Card Processing Centre",
        "200" => "Pick-up (general, no comments)",
        "201" => "Pick-up, expired card",
        "202" => "Pick-up, suspected fraud",
        "203" => "Pick-up, card acceptor contact card acquirer",
        "204" => "Pick-up, restricted card",
        "205" => "Pick-up, card acceptor call acquirer's security department",
        "206" => "Pick-up, allowable PIN tries exceeded",
        "207" => "Pick-up, special conditions",
        "208" => "Pick-up, lost card",
        "209" => "Pick-up, stolen card",
        "210" => "Pick-up, suspected counterfeit card",
        "300" => "Status message: file action successful",
        "301" => "Status message: file action not supported by receiver",
        "302" => "Status message: unable to locate record on file",
        "303" => "Status message: duplicate record, old record replaced",
        "304" => "Status message: file record field edit error",
        "305" => "Status message: file locked out",
        "306" => "Status message: file action not successful",
        "307" => "Status message: file data format error",
        "308" => "Status message: duplicate record, new record rejected",
        "309" => "Status message: unknown file",
        "400" => "Accepted (for reversal)",
        "500" => "Status message: reconciled, in balance",
        "501" => "Status message: reconciled, out of balance",
        "502" => "Status message: amount not reconciled, totals provided",
        "503" => "Status message: totals for reconciliation not available",
        "504" => "Status message: not reconciled, totals provided",
        "600" => "Accepted (for administrative info)",
        "601" => "Status message: impossible to trace back original transaction",
        "602" => "Status message: invalid transaction reference number",
        "603" => "Status message: reference number/PAN incompatible",
        "604" => "Status message: POS photograph is not available",
        "605" => "Status message: requested item supplied",
        "606" => "Status message: request cannot be fulfilled - required documentation is not available",
        "700" => "Accepted (for fee collection)",
        "800" => "Accepted (for network management)",
        "900" => "Advice acknowledged, no financial liability accepted",
        "901" => "Advice acknowledged, finansial liability accepted",
        "902" => "Decline reason message: invalid transaction",
        "903" => "Status message: re-enter transaction",
        "904" => "Decline reason message: format error",
        "905" => "Decline reason message: acqiurer not supported by switch",
        "906" => "Decline reason message: cutover in process",
        "907" => "Decline reason message: card issuer or switch inoperative",
        "908" => "Decline reason message: transaction destination cannot be found for routing",
        "909" => "Decline reason message: system malfunction",
        "910" => "Decline reason message: card issuer signed off",
        "911" => "Decline reason message: card issuer timed out",
        "912" => "Decline reason message: card issuer unavailable",
        "913" => "Decline reason message: duplicate transmission",
        "914" => "Decline reason message: not able to trace back to original transaction",
        "915" => "Decline reason message: reconciliation cutover or checkpoint error",
        "916" => "Decline reason message: MAC incorrect",
        "917" => "Decline reason message: MAC key sync error",
        "918" => "Decline reason message: no communication keys available for use",
        "919" => "Decline reason message: encryption key sync error",
        "920" => "Decline reason message: security software/hardware error - try again",
        "921" => "Decline reason message: security software/hardware error - no action",
        "922" => "Decline reason message: message number out of sequence",
        "923" => "Status message: request in progress",
        "940" => "Decline, blocked by fraud filter",
        "950" => "Decline reason message: violation of business arrangement"
      }

      attr_accessor :authorization, :error_code, :error_message, :success, :message

      def initialize params, options={}
        success = params[:result] == 'OK'
        message = RESPONSE_CODES[params[:result_code]]

        unless success
          error_code = params[:result_code]
          error_message = message
        end

        super success, message, params, options.merge(
          authorization: params[:transaction_id]
        )
      end

      def recurring?
        recc_pmnt_id.present? && recc_pmnt_expiry.present?
      end

      [:recc_pmnt_expiry, :recc_pmnt_id, :result, :result_code, :transaction_id].each do |name|
        define_method name do
          params[name]
        end
      end

      def _3d_secure
        params[:'3dsecure']
      end

      def result_message
        message
      end

      def [](value)
        params[value]
      end
    end
  end
end
